// @doist/todoist-api-typescript/dist/testUtils/testDefaults.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = exports.COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = exports.COMMENT_WITH_OPTIONALS_AS_NULL_TASK = exports.INVALID_COMMENT = exports.DEFAULT_COMMENT = exports.INVALID_ATTACHMENT = exports.DEFAULT_ATTACHMENT = exports.INVALID_USER = exports.DEFAULT_USER = exports.INVALID_LABEL = exports.DEFAULT_LABEL = exports.INVALID_SECTION = exports.DEFAULT_SECTION = exports.PROJECT_WITH_OPTIONALS_AS_NULL = exports.INVALID_PROJECT = exports.DEFAULT_PROJECT = exports.TASK_WITH_OPTIONALS_AS_NULL = exports.INVALID_TASK = exports.DEFAULT_TASK = exports.DEFAULT_QUICK_ADD_RESPONSE = exports.INVALID_DUE_DATE = exports.DEFAULT_DUE_DATE = exports.INVALID_ENTITY_ID = exports.DEFAULT_REQUEST_ID = exports.DEFAULT_AUTH_TOKEN = void 0;
var DEFAULT_TASK_ID = '1234';
var DEFAULT_TASK_CONTENT = 'This is a task';
var DEFAULT_TASK_DESCRIPTION = 'A description';
var DEFAULT_TASK_PRIORITY = 1;
var DEFAULT_ORDER = 3;
var DEFAULT_PROJECT_ID = '123';
var DEFAULT_PROJECT_NAME = 'This is a project';
var DEFAULT_PROJECT_VIEW_STYLE = 'list';
var DEFAULT_LABEL_ID = '456';
var DEFAULT_LABEL_NAME = 'This is a label';
var DEFAULT_SECTION_ID = '456';
var DEFAULT_SECTION_NAME = 'This is a section';
var DEFAULT_PARENT_ID = '5678';
var DEFAULT_ASSIGNEE = '1234';
var DEFAULT_CREATOR = '1234';
var DEFAULT_DATE = '2020-09-08T12:00:00Z';
var DEFAULT_ENTITY_COLOR = 'berry_red';
var DEFAULT_LABELS = ['personal', 'work', 'hobby'];
var DEFAULT_USER_ID = '5';
var DEFAULT_USER_NAME = 'A User';
var DEFAULT_USER_EMAIL = 'atestuser@doist.com';
var DEFAULT_COMMENT_ID = '4';
var DEFAULT_COMMENT_CONTENT = 'A comment';
exports.DEFAULT_AUTH_TOKEN = 'AToken';
exports.DEFAULT_REQUEST_ID = 'ARequestID';
exports.INVALID_ENTITY_ID = 1234;
exports.DEFAULT_DUE_DATE = {
    isRecurring: false,
    string: 'a date string',
    date: DEFAULT_DATE,
};
exports.INVALID_DUE_DATE = __assign(__assign({}, exports.DEFAULT_DUE_DATE), { isRecurring: 'false' });
exports.DEFAULT_QUICK_ADD_RESPONSE = {
    id: DEFAULT_TASK_ID,
    projectId: DEFAULT_PROJECT_ID,
    content: DEFAULT_TASK_CONTENT,
    description: DEFAULT_TASK_DESCRIPTION,
    priority: DEFAULT_TASK_PRIORITY,
    sectionId: DEFAULT_SECTION_ID,
    parentId: DEFAULT_PARENT_ID,
    childOrder: DEFAULT_ORDER,
    labels: DEFAULT_LABELS,
    responsibleUid: DEFAULT_ASSIGNEE,
    checked: false,
    addedAt: DEFAULT_DATE,
    addedByUid: DEFAULT_CREATOR,
    due: {
        date: DEFAULT_DATE,
        timezone: null,
        string: 'a date string',
        lang: 'en',
        isRecurring: false,
    },
};
exports.DEFAULT_TASK = {
    id: DEFAULT_TASK_ID,
    order: DEFAULT_ORDER,
    parentId: DEFAULT_PARENT_ID,
    content: DEFAULT_TASK_CONTENT,
    description: DEFAULT_TASK_DESCRIPTION,
    projectId: DEFAULT_PROJECT_ID,
    sectionId: DEFAULT_SECTION_ID,
    isCompleted: false,
    labels: DEFAULT_LABELS,
    priority: DEFAULT_TASK_PRIORITY,
    commentCount: 0,
    createdAt: DEFAULT_DATE,
    url: 'https://todoist.com/showTask?id=1234',
    due: exports.DEFAULT_DUE_DATE,
    assigneeId: DEFAULT_ASSIGNEE,
    creatorId: DEFAULT_CREATOR,
};
exports.INVALID_TASK = __assign(__assign({}, exports.DEFAULT_TASK), { due: exports.INVALID_DUE_DATE });
exports.TASK_WITH_OPTIONALS_AS_NULL = __assign(__assign({}, exports.DEFAULT_TASK), { due: null, assigneeId: null, assignerId: null, parentId: null, sectionId: null });
exports.DEFAULT_PROJECT = {
    id: DEFAULT_PROJECT_ID,
    name: DEFAULT_PROJECT_NAME,
    color: DEFAULT_ENTITY_COLOR,
    order: DEFAULT_ORDER,
    parentId: DEFAULT_PROJECT_ID,
    commentCount: 0,
    isFavorite: false,
    isShared: false,
    isInboxProject: false,
    isTeamInbox: false,
    viewStyle: DEFAULT_PROJECT_VIEW_STYLE,
    url: 'https://todoist.com/showProject?id=123',
};
exports.INVALID_PROJECT = __assign(__assign({}, exports.DEFAULT_PROJECT), { name: 123 });
exports.PROJECT_WITH_OPTIONALS_AS_NULL = __assign(__assign({}, exports.DEFAULT_PROJECT), { parentId: null });
exports.DEFAULT_SECTION = {
    id: DEFAULT_SECTION_ID,
    name: DEFAULT_SECTION_NAME,
    order: DEFAULT_ORDER,
    projectId: DEFAULT_PROJECT_ID,
};
exports.INVALID_SECTION = __assign(__assign({}, exports.DEFAULT_SECTION), { projectId: undefined });
exports.DEFAULT_LABEL = {
    id: DEFAULT_LABEL_ID,
    name: DEFAULT_LABEL_NAME,
    color: DEFAULT_ENTITY_COLOR,
    order: DEFAULT_ORDER,
    isFavorite: false,
};
exports.INVALID_LABEL = __assign(__assign({}, exports.DEFAULT_LABEL), { isFavorite: 'true' });
exports.DEFAULT_USER = {
    id: DEFAULT_USER_ID,
    name: DEFAULT_USER_NAME,
    email: DEFAULT_USER_EMAIL,
};
exports.INVALID_USER = __assign(__assign({}, exports.DEFAULT_USER), { email: undefined });
exports.DEFAULT_ATTACHMENT = {
    resourceType: 'file',
    fileType: 'image/png',
    fileUrl: 'https://someurl.com/image.jpg',
    uploadState: 'completed',
};
exports.INVALID_ATTACHMENT = __assign(__assign({}, exports.DEFAULT_ATTACHMENT), { uploadState: 'something random' });
exports.DEFAULT_COMMENT = {
    id: DEFAULT_COMMENT_ID,
    content: DEFAULT_COMMENT_CONTENT,
    projectId: DEFAULT_PROJECT_ID,
    attachment: exports.DEFAULT_ATTACHMENT,
    postedAt: DEFAULT_DATE,
};
exports.INVALID_COMMENT = __assign(__assign({}, exports.DEFAULT_COMMENT), { attachment: exports.INVALID_ATTACHMENT });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_TASK = __assign(__assign({}, exports.DEFAULT_COMMENT), { projectId: null, attachment: null });
exports.COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = __assign(__assign({}, exports.DEFAULT_COMMENT), { attachment: __assign(__assign({}, exports.DEFAULT_ATTACHMENT), { fileName: null, fileSize: null, fileType: null, fileDuration: null, uploadState: null, image: null, imageWidth: null, imageHeight: null, url: null, title: null }) });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = __assign(__assign({}, exports.DEFAULT_COMMENT), { taskId: null, attachment: null });
